crmprtd/insert.py

```diff
@@ -8,11 +8,217 @@
 speed and reliability. This phase is common to all networks.
 """
 
+from math import floor
+from sqlalchemy import and_
+import logging
+import time
+from sqlalchemy.exc import IntegrityError
+
+from crmprtd.db_exceptions import UniquenessError, InsertionError
+from pycds import Obs
+
+
+log = logging.getLogger('crmprtd.insert')
+
+
+class DBMetrics(object):
+    '''Keep track of database metrics during the insertion process.
+    '''
+
+    def __init__(self):
+        self.successes = 0
+        self.failures = 0
+
+    def clear(self):
+        self.successes = 0
+        self.failures = 0
+
+
+class Timer(object):
+    def __enter__(self):
+        self.start = time.time()
+        return self
+
+    def __exit__(self, *args):
+        self.end = time.time()
+        self.run_time = self.end - self.start
+
+
+def chunks(list, chunk_size):
+    for i in range(0, len(list), chunk_size):
+        yield list[i:i+chunk_size]
+
+
+def get_sample_indices(num_obs, sample_size):
+    if sample_size > num_obs:
+        sample_size = num_obs
+
+    index_buffer = floor(num_obs/sample_size)
+
+    sample_indices = [0]
+    index = 0
+    for i in range(sample_size - 1):
+        index += index_buffer
+        sample_indices.append(index)
+
+    return sample_indices
+
+
+def is_unique(sesh, history_id, vars_id, time):
+    q = sesh.query(Obs.id).filter(
+        and_(Obs.history_id == history_id, Obs.vars_id == vars_id,
+             Obs.time == time))
+    if q.count() > 0:
+        return False
+
+    return True
+
+
+def has_unique_obs(sesh, observations, sample_size):
+    sample_obs = []
+    sample_indices = get_sample_indices(len(observations), sample_size)
+
+    for index in sample_indices:
+        sample_obs.append(observations[index])
+
+    already_exists = 0
+    for sample in sample_obs:
+        if not is_unique(sesh, sample.history_id, sample.vars_id, sample.time):
+            already_exists += 1
```

Comment:
Rather than having a counter, and returning False if all of the pass, you can do the opposite: return True if the is_unique conditional passes. Then you don't need the counter. E.g.:

```python
for sample in sample_obs:
    if is_unique(...):
        return True
return False
```

Reply:
The reason for the counter is to see if **all** samples are duplicates.  The return value is used to determine whether to use the one-by-one or chunks+bisecting.  Do you want to keep that condition? Or if there are _any_ duplicates found in the samples we use the one-by-one strategy?

Reply:
Right, but logically for a given array of booleans `A`: `any(not(A)) == not(all(A))`. Maybe I didn't give you the right logic in my first comment, but my point is that if you get the logic right, it can be done w/o a counter.

Reply:
I think the name of the function is confusing me. What is the expected return value. True if it detects all duplicates and False otherwise? If so maybe the name should be `contains_all_duplicates`? Then I guess the logic would be:

```
for sample in sample_obs:
    if is_unique(...):
        return False
return True
```
?

Reply:
I think you are right.  I'll change some of the names to make things easier to understand.  I'm looking a it a day after I wrote it and I'm having trouble reading through it.